keanu-project/src/main/java/io/improbable/keanu/util/dot/WriteDot.java

```diff
@@ -0,0 +1,279 @@
+        verticesToProcessNow.add(vertex);
+
+        // Iterate over the graph till the specified degree or till there are no more unprocessed vertices left.
+        int iterationIndex = 0;
+        while(iterationIndex < degree && !verticesToProcessNow.isEmpty()) {
+
+            HashSet<Vertex> verticesToProcessNext = new HashSet<>();
+
+            for (Vertex v : verticesToProcessNow) {
+
+                // Use hashed IDs as the unique identifier for each vertex and store the other information in labels.
+                VertexId vId = v.getId();
+                setLabel(v);
+
+                // Iterate over all vertices connected to this one.
+                Stream<Vertex> connectedVertices = Stream.concat(v.getParents().stream(), v.getChildren().stream());
+                connectedVertices.forEach(connectedVertex -> {
+                    VertexId connectedVId = connectedVertex.getId();
+                    if (!processedVertices.contains(connectedVId)) {
+                        verticeIdsToConnectionInfo.put(concatenateVertexIds(connectedVId, vId), getEdgeString(connectedVId, vId));
+                        verticesToProcessNext.add(connectedVertex);
+                    }
+                });
+
+                // If there is any hyperparameter information to be added to edges, add it now.
+                DotAnnotation vertexAnnotation = v.getClass().getAnnotation(DotAnnotation.class);
+                if (vertexAnnotation != null && vertexAnnotation.displayHyperparameterInfo()) {
+                    applyHyperparameterInfo(v);
+                }
+
+                processedVertices.add(vId);
+            }
+
+            verticesToProcessNow = verticesToProcessNext;
+            iterationIndex++;
+        }
+
+        // Set labels for vertices that have not yet been processed.
+        for (Vertex v : verticesToProcessNow) {
+            setLabel(v);
+        }
+    }
+
+
+    // Utility function for creating a unique identifier for a connection between two vertices.
+    private static String concatenateVertexIds(VertexId v1, VertexId v2) {
+        if (v1.compareTo(v2) < 0) {
+            return v1.toString() + "_" + v2.toString();
+        }
+        else {
+            return v2.toString() + "_" + v1.toString();
+        }
+    }
+
+
+    // Utility function for forming a DOT string to represent an edge.
+    private static String getEdgeString(VertexId v1, VertexId v2) {
+        if (v1.compareTo(v2) < 0) {
+            return  "<" + v1.hashCode() + "> -> <" + v2.hashCode() + ">";
+        }
+        else {
+            return  "<" + v2.hashCode() + "> -> <" + v1.hashCode() + ">";
+        }
+    }
+
+
+    /**
+     * Creates and stores a label for the given vertex.
+     * Label contains the information that will be displayed for this vertex, such as vertex class or display name and value.
+     *
+     * @param v a vertex
+     */
+    private static void setLabel(Vertex v){
```

Comment:
This name could be confusing given it's shared with something on the Vertex class itself?

Reply:
It's a private method though, not intended to be used outside of this class. But I'll rename it to `setDotLabel`, as I guess it could be confusing if anyone wants to do more with this class.